tests/integration/synapseclient/core/test_download.py

```diff
@@ -343,6 +348,175 @@ async def test_download_expired_url(
+        project_model: Project,
+    ) -> None:
+        """Tests that a file download is retried with a presigned URL but it is expired."""
+        # GIVEN a file stored in synapse
+        original_file = utils.make_bogus_data_file(40000)
+        file = await File(path=original_file, parent_id=project_model.id).store_async(
+            synapse_client=syn
+        )
+
+        # AND extract the preSignedURL for that file
+        file_handle_response = get_file_handle_for_download(
+            file_handle_id=file.data_file_handle_id,
+            synapse_id=file.id,
+            entity_type="FileEntity",
+            synapse_client=syn,
+        )
+        url = file_handle_response["preSignedURL"]
+        expired_url = _expire_url(url)
+
+        # get_file_handle_for_download only runs if we are handling an expired URL and not
+        spy_file_handle = mocker.spy(download_functions, "get_file_handle_for_download")
+
+        # WHEN I download the file with a simulated expired url
+        with pytest.raises(
+            SynapseError,
+            match="The provided pre-signed URL has expired. Please provide a new pre-signed URL.",
+        ):
+            path = download_from_url(
+                url=expired_url,
+                destination=tempfile.gettempdir(),
+                entity_id=None,
+                file_handle_associate_type=None,
+                file_handle_id=None,
+                expected_md5=file.file_handle.content_md5,
+                url_is_presigned=True,
+            )
+
+        # THEN the expired URL is refreshed
+        spy_file_handle.assert_not_called()
+
+    async def test_download_via_presigned_url_md5_matches(
+        self,
+        syn: Synapse,
+        mocker: MockerFixture,
+        project_model: Project,
+    ) -> None:
+        """Tests that a file download is retried with a presigned URL when the md5 matches."""
+        # GIVEN a file stored in synapse
+        original_file = utils.make_bogus_data_file(40000)
+        file = await File(path=original_file, parent_id=project_model.id).store_async(
+            synapse_client=syn
+        )
+
+        # AND extract the preSignedURL for that file
+        file_handle_response = get_file_handle_for_download(
+            file_handle_id=file.data_file_handle_id,
+            synapse_id=file.id,
+            entity_type="FileEntity",
+            synapse_client=syn,
+        )
+        url = file_handle_response["preSignedURL"]
+
+        # get_file_handle_for_download only runs if we are handling an expired URL and not
+        spy_file_handle = mocker.spy(download_functions, "get_file_handle_for_download")
+
+        path = download_from_url(
+            url=url,
+            destination=tempfile.gettempdir(),
+            entity_id=None,
+            file_handle_associate_type=None,
+            file_handle_id=None,
+            expected_md5=file.file_handle.content_md5,
+            url_is_presigned=True,
+        )
```

Comment:
Is the intention that these function calls be made for FileEntities stored on Synapse (Or is this just that your testing this case?)

The downside of directly calling this function for Synapse FileEntities is that it bypasses the cache mechanism set up.

If the intention is that "Give a random presigned URL, just download the thing", then this is a fine solution